tests/msc2716_test.go

```diff
@@ -0,0 +1,643 @@
+		t,
+		"POST",
+		[]string{"_matrix", "client", "r0", "register"},
+		client.WithRawBody(json.RawMessage(fmt.Sprintf(`{ "type": "m.login.application_service", "username": "%s" }`, virtualUserLocalpart))),
+		client.WithContentType("application/json"),
+	)
+
+	if res.StatusCode == 200 {
+		return
+	}
+
+	body := client.ParseJSON(t, res)
+	errcode := client.GetJSONFieldStr(t, body, "errcode")
+
+	if res.StatusCode == 400 && errcode == "M_USER_IN_USE" {
+		return
+	} else {
+		errorMessage := client.GetJSONFieldStr(t, body, "error")
+		t.Fatalf("msc2716.ensureVirtualUserRegistered failed to register: (%s) %s", errcode, errorMessage)
+	}
+}
+
+func createMessagesInRoom(t *testing.T, c *client.CSAPI, roomID string, count int) []string {
+	evs := make([]string, count)
+	for i := 0; i < len(evs); i++ {
+		newEvent := b.Event{
+			Type: "m.room.message",
+			Content: map[string]interface{}{
+				"msgtype": "m.text",
+				"body":    fmt.Sprintf("Message %d", i),
+			},
+		}
+		newEventId := c.SendEventSynced(t, roomID, newEvent)
+		evs[i] = newEventId
+	}
+
+	return evs
+}
+
+var chunkCount int64 = 0
+
+func backfillBatchHistoricalMessages(
+	t *testing.T,
+	c *client.CSAPI,
+	virtualUserID string,
+	roomID string,
+	insertAfterEventId string,
+	insertTime time.Time,
+	chunkID string,
+	count int,
+	status int,
+) (res *http.Response) {
+	// Timestamp in milliseconds
+	insertOriginServerTs := uint64(insertTime.UnixNano() / int64(time.Millisecond))
+
+	timeBetweenMessagesMS := uint64(TimeBetweenMessages / time.Millisecond)
+
+	evs := make([]map[string]interface{}, count)
+	for i := 0; i < len(evs); i++ {
+		newEvent := map[string]interface{}{
+			"type":             "m.room.message",
+			"sender":           virtualUserID,
+			"origin_server_ts": insertOriginServerTs + (timeBetweenMessagesMS * uint64(i)),
+			"content": map[string]interface{}{
+				"msgtype":          "m.text",
+				"body":             fmt.Sprintf("Historical %d (chunk=%d)", i, chunkCount),
+				MSC2716_HISTORICAL: true,
+			},
+		}
+
+		evs[i] = newEvent
+	}
+
+	joinEvent := map[string]interface{}{
+		"type":             "m.room.member",
+		"sender":           virtualUserID,
+		"origin_server_ts": insertOriginServerTs,
+		"content": map[string]interface{}{
+			"membership": "join",
+		},
+		"state_key": virtualUserID,
+	}
+
+	query := make(url.Values, 2)
+	query.Add("prev_event", insertAfterEventId)
+	query.Add("user_id", virtualUserID)
+	// If provided, connect the chunk to the last insertion point
+	if chunkID != "" {
+		query.Add("chunk_id", chunkID)
+	}
+
+	b, err := json.Marshal(map[string]interface{}{
+		"events":                evs,
+		"state_events_at_start": []map[string]interface{}{joinEvent},
+	})
+	if err != nil {
+		t.Fatalf("msc2716.backfillBatchHistoricalMessages failed to marshal JSON body: %s", err)
+	}
+
+	res = c.DoFunc(
+		t,
+		"POST",
+		[]string{"_matrix", "client", "unstable", "org.matrix.msc2716", "rooms", roomID, "batch_send"},
+		client.WithRawBody(b),
```

Comment:
Use `client.WithJSONBody` so you don't need to `json.Marshal` on `:592` (you can inline it).